### PR TITLE
list: refactor command options & move --unbrewed switch

### PR DESCRIPTION
**Pull request by EricFromCanada**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?
- [x] Have you successfully run `brew man` locally and committed any changes?

-----

Rearrange the logic so this command behaves more predictably and errors out where it should. Also move the `--unbrewed` switch to `brew --prefix` which seems like a better fit. It now operates in one of these modes:
- (no named args): list all installed formulae/casks, in columns or a list
- (with named args and none of the options below): list files or artifacts for specified formulae/casks
- `--full-name`: show full names of formulae/casks, either all installed or only those specified, in columns or a list
- `--versions`: list installed or specified formulae/casks with their version numbers
- `--pinned`: only list any or specified pinned formulae

### CODE REVIEW

File: Library/Homebrew/cmd/--prefix.rb [hunk 1]
```diff
@@ -20,18 +20,67 @@ def __prefix_args
         If <formula> is provided, display the location in the Cellar where <formula>
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/--prefix.rb:20]: On lines 17-18 it'd be nice to use https://github.com/Homebrew/brew/blob/486e76c1f2bc8e350be26000be9f7d770efa9a30/Library/Homebrew/global.rb#L71-L73

File: Library/Homebrew/cmd/list.rb [hunk 2]
```diff
@@ -69,45 +75,52 @@ def list_args
   def list
     args = list_args.parse
 
-    return list_casks(args: args) if args.cask?
-
     if args.unbrewed?
       raise UsageError, "`--unbrewed` does not take a formula/cask argument." unless args.no_named?
 
-      return list_unbrewed
+      odeprecated "`brew list --unbrewed`", "`brew --prefix --unbrewed`"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/list.rb:81]:
```suggestion
      odisabled "`brew list --unbrewed`", "`brew --prefix --unbrewed`"
```

unless this logic is passed through to `brew --prefix --unbrewed`.

File: Library/Homebrew/cmd/list.rb [hunk 3]
```diff
@@ -20,26 +20,26 @@ def list_args
...
+        If <cask> is provided, list its artifacts.
       EOS
       switch "--formula", "--formulae",
-             description: "List only formulae."
+             description: "List only formulae, or treat all named arguments as formulae."
       switch "--cask", "--casks",
-             description: "List only casks, or <cask> if provided."
+             description: "List only casks, or treat all named arguments as casks."
       switch "--unbrewed",
              description: "List files in Homebrew's prefix not installed by Homebrew."
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/list.rb:30]:
```suggestion
      switch "--unbrewed",
             description: "List files in Homebrew's prefix not installed by Homebrew."
```

I wonder if a `deprecated` flag or similar could:
- set `odeprecated` if this flag is called
- stop it from being shown in the manpage and help.

Comment (EricFromCanada) [reply Library/Homebrew/cmd/list.rb]: I added a `replacement:` option to `switch()` to pass a suggestion to `odisabled` if the switch is used, but I wasn't able to hide it from the `--help` output since that comes directly from `OptionParser`.

File: Library/Homebrew/cmd/list.rb [hunk 4]
```diff
@@ -69,45 +76,46 @@ def list_args
...
 
-    if args.pinned? || args.versions?
+    if args.full_name?
+      unless args.cask?
+        formula_names = args.no_named? ? Formula.installed : args.named.to_resolved_formulae
+        full_formula_names = formula_names.map(&:full_name).sort(&tap_and_name_comparison)
+        full_formula_names = Formatter.columns(full_formula_names) unless args.public_send(:'1?')
+        puts full_formula_names unless full_formula_names.blank?
+      end
+      unless args.formula? || (!args.no_named? && !args.cask?)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/list.rb:94]: @EricFromCanada Can this be refactored to not use `unless ... ||` which is hard to follow.

File: Library/Homebrew/cmd/list.rb [hunk 5]
```diff
@@ -69,45 +76,46 @@ def list_args
...
+        cask_names =
+          args.no_named? ? Cask::Caskroom.casks : args.named.to_formulae_and_casks(only: :cask, method: :resolve)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/list.rb:96]:
```suggestion
        cask_names = if args.no_named?
          Cask::Caskroom.casks
        else
          args.named.to_formulae_and_casks(only: :cask, method: :resolve)
        end
```

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks @EricFromCanada!
